torchsde/_core/adjoint_sde.py

```diff
@@ -0,0 +1,425 @@
+        y, adj_y = y_aug[:n_tensors], y_aug[n_tensors:2 * n_tensors]
+
+        with torch.enable_grad():
+            y = [y_.detach().requires_grad_(True) for y_ in y]
+            minus_adj_y = [-adj_y_.detach() for adj_y_ in adj_y]
+            minus_f = [-f_ for f_ in sde.f(-t, y)]
+            vjp_y_and_params = misc.grad(
+                outputs=minus_f,
+                inputs=y + params,
+                grad_outputs=minus_adj_y,
+                allow_unused=True,
+            )
+            vjp_y, vjp_params = vjp_y_and_params[:n_tensors], vjp_y_and_params[n_tensors:]
+            vjp_params = misc.flatten(vjp_params)
+
+        return (*minus_f, *vjp_y, vjp_params)
+
+    def f_corrected_default(self, t, y_aug):  # For Ito general/scalar.
+        raise NotImplementedError
+
+    def f_corrected_diagonal(self, t, y_aug):  # For Ito diagonal.
+        sde, params, n_tensors = self._base_sde, self._params, self._n_tensors
+        y, adj_y = y_aug[:n_tensors], y_aug[n_tensors:2 * n_tensors]
+
+        with torch.enable_grad():
+            y = [y_.detach().requires_grad_(True) for y_ in y]
+            adj_y = [adj_y_.detach() for adj_y_ in adj_y]
+
+            g_eval = sde.g(-t, y)
+            gdg = misc.grad(
+                outputs=g_eval,
+                inputs=y,
+                grad_outputs=g_eval,
+                allow_unused=True,
+                create_graph=True
+            )
+            f_eval = sde.f(-t, y)
+            # Stratonovich correction for reverse-time.
+            f_eval_corrected = misc.seq_sub(gdg, f_eval)
+            vjp_y_and_params = misc.grad(
+                outputs=f_eval_corrected,
+                inputs=y + params,
+                grad_outputs=[-adj_y_ for adj_y_ in adj_y],
+                allow_unused=True,
+                create_graph=True
+            )
+            vjp_y, vjp_params = vjp_y_and_params[:n_tensors], vjp_y_and_params[n_tensors:]
+            vjp_params = misc.flatten(vjp_params)
+
+            adj_times_dgdx = misc.grad(
+                outputs=g_eval,
+                inputs=y,
+                grad_outputs=adj_y,
+                allow_unused=True,
+                create_graph=True
+            )
+
+            # Converting the *adjoint* Stratonovich backward SDE to Itô.
+            extra_vjp_y_and_params = misc.grad(
+                outputs=g_eval,
+                inputs=y + params,
+                grad_outputs=adj_times_dgdx,
+                allow_unused=True,
+            )
+            extra_vjp_y, extra_vjp_params = extra_vjp_y_and_params[:n_tensors], extra_vjp_y_and_params[n_tensors:]
+            extra_vjp_params = misc.flatten(extra_vjp_params)
+
+            vjp_y = misc.seq_add(vjp_y, extra_vjp_y)
+            vjp_params = vjp_params + extra_vjp_params
+
+        return (*f_eval_corrected, *vjp_y, vjp_params)
+
+    ########################################
+    #                g_prod                #
+    ########################################
+
```

Comment:
Yeah, I am completely aware of the progress in #31, but I need to think about this more. There's a reason that the current API uses `g_prod` as opposed to `g`, and that reason is mostly related to the adjoint computation. It's more consistent this way, since directly playing with `g` for the adjoint make likely clutter other places.

Reply:
Well, the point is that the new Milstein solvers expect to have `prod` available to call.
With both this and the above comment, if I get some time I'll try and understand the adjoint a bit better to see if I can help out.

Reply:
Obviously, for the adjoint, `g_prod` is a vjp, so it's not just eval g, eval v, then multiply. There's a reason I made this `g_prod` function, and it's not like I'm a boor and I want to make things inefficient. 

At this point, I'm not interested in premature optimizations that would scatter the codebase. We could and should do the optimizations, but it should be at a later time with a clear plan.